Integrate with Bugsnag

ermshiperete opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sillsdev/web-languageforge/227)
<!-- Reviewable:end -->

irahopkinson commented:
I realize this is WIP but thought I would add comments since I've read through it all anyway.

---

Reviewed 14 of 14 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[src/composer.json, line 34 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LAud1-O0_T-sleIZnVa:-LAud1-O0_T-sleIZnVb:bbwonn3) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/composer.json#L34)):*
> ```JSON
>     "twilio/sdk": "dev-master",
>     "firebase/php-jwt": "^5.0",
>     "bugsnag/bugsnag-silex": "^2.0"
> ```

I would prefer these in alphabetical order. I realize `firebase` is already in the wrong place and needs to moved too.

---

*[src/index.php, line 26 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LAudWhJ3hF_ZAFPhsL9:-LAudWhJ3hF_ZAFPhsLA:bwumvas) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/index.php#L26)):*
> ```PHP
> */
> 
> BugsnagExceptionHandler::setup($app, 'd7dbf8324db4b392623f69bd0f2098a6');
> ```

Doesn't this line need to go after the comment section rather than in the middle of it?
Does it contain a secret key?
Also I would prefer the original indenting to match with PHP elsewhere.

---

*[src/angular-app/bellows/core/core.module.ts, line 3 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LAueHAEAhEt5yUQhVre:-LAueHAEAhEt5yUQhVrf:bb0c0if) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/angular-app/bellows/core/core.module.ts#L3)):*
> ```XML
> import * as angular from 'angular';
> 
> import { ActivityService } from './api/activity.service';
> ```

You don't need to add the spaces in everywhere. We are changing back to having no spaces.

---

*[src/angular-app/bellows/core/exception-handling.service.ts, line 3 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LAuf5Hs2xzHJbl9IDKb:-LAuf5Hs2xzHJbl9IDKc:bol1p19) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/angular-app/bellows/core/exception-handling.service.ts#L3)):*
> ```XML
> import * as angular from 'angular';
> import Bugsnag from 'bugsnag-js';
> import { Client } from 'bugsnag-js/types/client';
> ```

Style: put a blank line in to separate vendor imports from local code imports.

---

*[src/angular-app/bellows/core/exception-handling.service.ts, line 9 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LAufRui2MWGZ383sCn3:-LAufRui2MWGZ383sCn4:b-8ut6au) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/angular-app/bellows/core/exception-handling.service.ts#L9)):*
> ```XML
> 
> export class ExceptionHandlingService {
>   static $inject: string[] = ['$log', 'userService', 'sessionService'];
> ```

Style: place the `$inject` definition directly above the constructor without a blank line in between. The two are intimately linked. I also align the CR between the two.

---

*[src/angular-app/bellows/core/exception-handling.service.ts, line 15 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LAuhUZHEbZD2oD_7dtb:-LAuhUZHEbZD2oD_7dtc:b-47vpfi) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/angular-app/bellows/core/exception-handling.service.ts#L15)):*
> ```XML
>   constructor(private $log: angular.ILogService, private userService: UserService,
>               private sessionService: SessionService) {
>     this.bugsnagClient = Bugsnag({ apiKey: '0ac621681992952c356d54bc01d529b1' }); // d7dbf8324db4b392623f69bd0f2098a6
> ```

Is the `apiKey` meant to be a secret?

---

*[src/angular-app/bellows/core/exception-handling.service.ts, line 66 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LAuiCmJ7r1Q6j1foBnE:-LAuiCmJ7r1Q6j1foBnF:bffps1r) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/angular-app/bellows/core/exception-handling.service.ts#L66)):*
<details><summary><i alt="summary">Quoted 6 lines of code…</i></summary>

> ```
>   .factory('$exceptionHandler', ($log: angular.ILogService, userService: UserService,
>                                  sessionService: SessionService) => {
>     const handler = new ExceptionHandlingService($log, userService, sessionService);
>  
>     return (exception: Error, cause?: string) => handler.reportError(exception, cause);
>   })
> ```
</details>

Shouldn't need this line and is causing your error.

---

*[src/angular-app/bellows/core/api/json-rpc.service.ts, line 28 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LAuiONaB4X-TP0hbY2t:-LAuiONaB4X-TP0hbY2u:btk96p8) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/angular-app/bellows/core/api/json-rpc.service.ts#L28)):*
> ```XML
> 
>   static $inject: string[] = ['$http', '$window', 'error', 'exceptionHandler'];
> ```

Style: remove this blank line. `$inject` and the constructor are linked.

---

*[src/Api/Library/Shared/Palaso/JsonRpcServer.php, line 33 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LAuiug806zxTJ7AyI3H:-LAuiug806zxTJ7AyI3I:bnbvkj6) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/Api/Library/Shared/Palaso/JsonRpcServer.php#L33)):*
> ```PHP
> use Palaso\Utilities\CodeGuard;
> use Symfony\Component\HttpFoundation\Request;
> use Bugsnag\Silex\Silex1ServiceProvider;
> ```

Order alphabetically.

---

*[src/Api/Library/Shared/Palaso/Exception/BugsnagExceptionHandler.php, line 12 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LAunQaZE3jmcemOOO-v:-LAunQaZE3jmcemOOO-w:b-a0ohj1) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/Api/Library/Shared/Palaso/Exception/BugsnagExceptionHandler.php#L12)):*
> ```
> require_once $appPath . '/vendor/autoload.php';
> require_once $appPath . '/config.php';
> require_once $appPath . '/version.php';
> ```
Usually these are not needed if this class is called inside the application. I can't see where that is not the case?

---


*Comments from [Reviewable](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-:-LAuob0D-y-DOncZuj-s:b-t852x4)*
<!-- Sent from Reviewable.io -->

megahirt commented:
I've added comments as well.  Nice work @ermshiperete on getting your hands dirty with PHP!

---

Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[src/index.php, line 26 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LAudWhJ3hF_ZAFPhsL9:-LAv0YZKBzXWmGzdabHn:b-1mqxso) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/index.php#L26)):*
<details><summary><i>Previously, irahopkinson (Ira Hopkinson) wrote…</i></summary><blockquote>

Doesn't this line need to go after the comment section rather than in the middle of it?
Does it contain a secret key?
Also I would prefer the original indenting to match with PHP elsewhere.
</blockquote></details>

I don't really like the key being in our repo either.  I know it's mostly harmless though.  @ermshiperete can we stamp it at build time?

---

*[src/angular-app/bellows/core/exception-handling.service.ts, line 66 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LAuiCmJ7r1Q6j1foBnE:-LAv1Q8pEXPt9erEPSpk:b590xi7) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/angular-app/bellows/core/exception-handling.service.ts#L66)):*
<details><summary><i>Previously, irahopkinson (Ira Hopkinson) wrote…</i></summary><blockquote>

> ```
>   .factory('$exceptionHandler', ($log: angular.ILogService, userService: UserService,
>                                  sessionService: SessionService) => {
>     const handler = new ExceptionHandlingService($log, userService, sessionService);
>  
>     return (exception: Error, cause?: string) => handler.reportError(exception, cause);
>   })
> ```
Shouldn't need this line and is causing your error.
</blockquote></details>

@irahopkinson I couldn't figure out which line you said is not necessary from your quote.

---

*[src/Api/Library/Shared/Palaso/JsonRpcServer.php, line 53 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LAv3UxW7IPmC5G7-BCD:-LAv3UxW7IPmC5G7-BCE:b-4ux8t0) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/Api/Library/Shared/Palaso/JsonRpcServer.php#L53)):*
> ```PHP
>      * @throws \Exception
>      */
>     public static function handle(Request $request, $api) {
> ```

I think we should change the signature of handle() to accept an "Application $app", and then pass that into bugsnag.

---

*[src/Api/Library/Shared/Palaso/JsonRpcServer.php, line 114 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LAv2BKVAO8LwtGaVxDY:-LAv2BKVAO8LwtGaVxDZ:b-7s4s80) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/Api/Library/Shared/Palaso/JsonRpcServer.php#L114)):*
> ```PHP
>             }
>             else {
>                 BugsnagExceptionHandler::getBugsnag()->notifyException($e);
> ```

cool.

---

*[src/Api/Library/Shared/Palaso/Exception/BugsnagExceptionHandler.php, line 12 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LAunQaZE3jmcemOOO-v:-LAv2XxY7N3Hnl5ihCX4:b8kt7lt) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/Api/Library/Shared/Palaso/Exception/BugsnagExceptionHandler.php#L12)):*
<details><summary><i>Previously, irahopkinson (Ira Hopkinson) wrote…</i></summary><blockquote>

> ```
> require_once $appPath . '/vendor/autoload.php';
> require_once $appPath . '/config.php';
> require_once $appPath . '/version.php';
> ```
Usually these are not needed if this class is called inside the application. I can't see where that is not the case?
</blockquote></details>

Agreed.  This code looks like it is bootstrapping an additional instance of the Silex Application.  Instead, we should modify the signature of the JsonRpcServer::handle() to take the $app parameter.  From $app you can get project and user information, as you have done before.

We don't want any "require_once" in our classes.

---

*[src/Api/Library/Shared/Palaso/Exception/BugsnagExceptionHandler.php, line 44 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LAv2zQ60auErG4gID5t:-LAv2zQ60auErG4gID5u:b-6hlvar) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/Api/Library/Shared/Palaso/Exception/BugsnagExceptionHandler.php#L44)):*
> ```PHP
>                 $report->setMetaData([
>                     'app' => [
>                         'projectId' => $application['session']->get('projectId')
> ```

Let's set the projectId, projectName and projectCode in the metadata as well.  You can do $project = new ProjectModel($projectId) and get the information from $project that way.

---


*Comments from [Reviewable](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-:-LAv3fjS8Zij4goiZ7wh:b-5jmxie)*
<!-- Sent from Reviewable.io -->

irahopkinson commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[src/angular-app/bellows/core/exception-handling.service.ts, line 66 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LAuiCmJ7r1Q6j1foBnE:-LAvQmP6EvRM7JVDii5o:b1zenfk) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/angular-app/bellows/core/exception-handling.service.ts#L66)):*
<details><summary><i>Previously, megahirt (Chris Hirt) wrote…</i></summary><blockquote>

@irahopkinson I couldn't figure out which line you said is not necessary from your quote.
</blockquote></details>

The entire `.factory(...)` line that was quoted.

---


*Comments from [Reviewable](https://reviewable.io/reviews/sillsdev/web-languageforge/227)*
<!-- Sent from Reviewable.io -->

ermshiperete commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[src/composer.json, line 34 at r1](https://beta.reviewable.io/reviews/sillsdev/web-languageforge/227#-LAud1-O0_T-sleIZnVa:-LAvUgbM7TPaGcor_EhC:b-896fix) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/composer.json#L34)):*
<details><summary><i>Previously, irahopkinson (Ira Hopkinson) wrote…</i></summary><blockquote>

I would prefer these in alphabetical order. I realize `firebase` is already in the wrong place and needs to moved too.
</blockquote></details>

Done.

---

*[src/index.php, line 26 at r1](https://beta.reviewable.io/reviews/sillsdev/web-languageforge/227#-LAudWhJ3hF_ZAFPhsL9:-LAvV2kU0Dhju2Rr86gy:b-uwfdzm) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/index.php#L26)):*
<details><summary><i>Previously, megahirt (Chris Hirt) wrote…</i></summary><blockquote>

I don't really like the key being in our repo either.  I know it's mostly harmless though.  @ermshiperete can we stamp it at build time?
</blockquote></details>

Not having the key in the code is one of the things I still wanted to do. This also needs to be a different key for LF and SF.

I put the code above the comment because the setup for bugsnag wil be identical regardless of the environment, but I can move it down if you prefer that.

---

*[src/angular-app/bellows/core/core.module.ts, line 3 at r1](https://beta.reviewable.io/reviews/sillsdev/web-languageforge/227#-LAueHAEAhEt5yUQhVre:-LAvWNBzDpcSYX7rWybz:b-lugh1o) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/angular-app/bellows/core/core.module.ts#L3)):*
<details><summary><i>Previously, irahopkinson (Ira Hopkinson) wrote…</i></summary><blockquote>

You don't need to add the spaces in everywhere. We are changing back to having no spaces.
</blockquote></details>

That was VS Code that did that... Done.

---

*[src/angular-app/bellows/core/exception-handling.service.ts, line 3 at r1](https://beta.reviewable.io/reviews/sillsdev/web-languageforge/227#-LAuf5Hs2xzHJbl9IDKb:-LAvXP8zE1gNDnY-5GnT:b-896fix) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/angular-app/bellows/core/exception-handling.service.ts#L3)):*
<details><summary><i>Previously, irahopkinson (Ira Hopkinson) wrote…</i></summary><blockquote>

Style: put a blank line in to separate vendor imports from local code imports.
</blockquote></details>

Done.

---

*[src/angular-app/bellows/core/exception-handling.service.ts, line 9 at r1](https://beta.reviewable.io/reviews/sillsdev/web-languageforge/227#-LAufRui2MWGZ383sCn3:-LAvXW-G7aTbUkASN5yy:b-wvwja3) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/angular-app/bellows/core/exception-handling.service.ts#L9)):*
<details><summary><i>Previously, irahopkinson (Ira Hopkinson) wrote…</i></summary><blockquote>

Style: place the `$inject` definition directly above the constructor without a blank line in between. The two are intimately linked. I also align the CR between the two.
</blockquote></details>

What's CR?

---

*[src/angular-app/bellows/core/exception-handling.service.ts, line 15 at r1](https://beta.reviewable.io/reviews/sillsdev/web-languageforge/227#-LAuhUZHEbZD2oD_7dtb:-LAvXbh01aeAZTJRPunY:b-85lph4) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/angular-app/bellows/core/exception-handling.service.ts#L15)):*
<details><summary><i>Previously, irahopkinson (Ira Hopkinson) wrote…</i></summary><blockquote>

Is the `apiKey` meant to be a secret?
</blockquote></details>

Not finished yet :smile: 

---

*[src/angular-app/bellows/core/api/json-rpc.service.ts, line 28 at r1](https://beta.reviewable.io/reviews/sillsdev/web-languageforge/227#-LAuiONaB4X-TP0hbY2t:-LAvY0DC2wAHLHqGet0T:b-896fix) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/angular-app/bellows/core/api/json-rpc.service.ts#L28)):*
<details><summary><i>Previously, irahopkinson (Ira Hopkinson) wrote…</i></summary><blockquote>

Style: remove this blank line. `$inject` and the constructor are linked.
</blockquote></details>

Done.

---

*[src/Api/Library/Shared/Palaso/JsonRpcServer.php, line 33 at r1](https://beta.reviewable.io/reviews/sillsdev/web-languageforge/227#-LAuiug806zxTJ7AyI3H:-LAvYQ0pEkE6JvinIGZ0:b-fsvoz2) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/Api/Library/Shared/Palaso/JsonRpcServer.php#L33)):*
<details><summary><i>Previously, irahopkinson (Ira Hopkinson) wrote…</i></summary><blockquote>

Order alphabetically.
</blockquote></details>

Turns out that line isn't needed at all.

---

*[src/Api/Library/Shared/Palaso/JsonRpcServer.php, line 53 at r1](https://beta.reviewable.io/reviews/sillsdev/web-languageforge/227#-LAv3UxW7IPmC5G7-BCD:-LAw2ba2719xuW1Z5yqc:b-896fix) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/Api/Library/Shared/Palaso/JsonRpcServer.php#L53)):*
<details><summary><i>Previously, megahirt (Chris Hirt) wrote…</i></summary><blockquote>

I think we should change the signature of handle() to accept an "Application $app", and then pass that into bugsnag.
</blockquote></details>

Done.

---

*[src/Api/Library/Shared/Palaso/Exception/BugsnagExceptionHandler.php, line 12 at r1](https://beta.reviewable.io/reviews/sillsdev/web-languageforge/227#-LAunQaZE3jmcemOOO-v:-LAvZG2M4fkWogPXHmTP:brai4oq) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/Api/Library/Shared/Palaso/Exception/BugsnagExceptionHandler.php#L12)):*
<details><summary><i>Previously, megahirt (Chris Hirt) wrote…</i></summary><blockquote>

Agreed.  This code looks like it is bootstrapping an additional instance of the Silex Application.  Instead, we should modify the signature of the JsonRpcServer::handle() to take the $app parameter.  From $app you can get project and user information, as you have done before.

We don't want any "require_once" in our classes.
</blockquote></details>

Done.

But then I'm missing something else. With these three lines (plus the `$appPath =  realpath...`) removed, I get an error:

```
Notice: Use of undefined constant VERSION - assumed 'VERSION' in 
languageforge/src/Api/Library/Shared/Palaso/Exception/BugsnagExceptionHandler.php on line 24
Call Stack
#	Time	Memory	Function	Location
1	0.0074	361232	{main}( )	.../index.php:0
2	0.0654	395112	Api\Library\Shared\Palaso\Exception\BugsnagExceptionHandler::setup( ???, ??? )	.../index.php:26
```

So it seems I need the `require_once $appPath . '/version.php';` line.

---

*[src/Api/Library/Shared/Palaso/Exception/BugsnagExceptionHandler.php, line 44 at r1](https://beta.reviewable.io/reviews/sillsdev/web-languageforge/227#-LAv2zQ60auErG4gID5t:-LAvZOfSBlL8016DX3hQ:bkyrvmg) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/Api/Library/Shared/Palaso/Exception/BugsnagExceptionHandler.php#L44)):*
<details><summary><i>Previously, megahirt (Chris Hirt) wrote…</i></summary><blockquote>

Let's set the projectId, projectName and projectCode in the metadata as well.  You can do $project = new ProjectModel($projectId) and get the information from $project that way.
</blockquote></details>

Yes. I noticed that the projectId isn't very useful when I looked at the entries that got created on Bugsnag.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/sillsdev/web-languageforge/227)*
<!-- Sent from Reviewable.io -->

irahopkinson commented:
Reviewed 9 of 9 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[src/index.php, line 26 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LAudWhJ3hF_ZAFPhsL9:-LAy_2ah8ow_IaTy9w9m:brktnbf) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/index.php#L26)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Not having the key in the code is one of the things I still wanted to do. This also needs to be a different key for LF and SF.

I put the code above the comment because the setup for bugsnag wil be identical regardless of the environment, but I can move it down if you prefer that.
</blockquote></details>

That's reasonable; leave it there.

---

*[src/angular-app/bellows/core/exception-handling.service.ts, line 3 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LAuf5Hs2xzHJbl9IDKb:-LAy_z49DfC4hq99CM4P:btw7tgt) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/angular-app/bellows/core/exception-handling.service.ts#L3)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Umm, `Client` is vendor code, move the blank line down one more.

---

*[src/angular-app/bellows/core/exception-handling.service.ts, line 9 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LAufRui2MWGZ383sCn3:-LAy__uA-vL9YsEtfPxU:b-v89b6x) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/angular-app/bellows/core/exception-handling.service.ts#L9)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

What's CR?
</blockquote></details>

Carriage Return

---


*Comments from [Reviewable](https://reviewable.io/reviews/sillsdev/web-languageforge/227)*
<!-- Sent from Reviewable.io -->

irahopkinson commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[src/angular-app/bellows/core/exception-handling.service.ts, line 61 at r2](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LB-DH-uCQxGluIcmBZK:-LB-DH-uCQxGluIcmBZL:bcujn44) ([raw file](https://github.com/sillsdev/web-languageforge/blob/848b3151898c779309ece2ef8dc81f635900545c/src/angular-app/bellows/core/exception-handling.service.ts#L61)):*
> ```XML
> export const ExceptionOverrideModule = angular
>   .module('exceptionOverride', [ ])
>   .service('exceptionHandler', ExceptionHandlingService)
> ```

This service is not doing anything and can be removed (since you are using the factory below).

---

*[src/angular-app/bellows/core/api/json-rpc.service.ts, line 29 at r2](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LB-0vdhC3F6gZWhfGll:-LB-0vdhC3F6gZWhfGlm:blcydkb) ([raw file](https://github.com/sillsdev/web-languageforge/blob/848b3151898c779309ece2ef8dc81f635900545c/src/angular-app/bellows/core/api/json-rpc.service.ts#L29)):*
> ```XML
>   static $inject: string[] = ['$http', '$window', 'error', 'exceptionHandler'];
>   constructor(private $http: angular.IHttpService, private $window: angular.IWindowService, private error: ErrorService,
>               private exceptionHandler: ExceptionHandlingService) {
> ```

I'm wondering why the `ExceptionHandlingService` is injected here since it is never used? Pretty sure it can be removed.

---

*[src/angular-app/bellows/core/api/json-rpc.service.ts, line 148 at r2](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LB-DvVW6IGxUREYGCgt:-LB-DvVW6IGxUREYGCgu:bixvw9c) ([raw file](https://github.com/sillsdev/web-languageforge/blob/848b3151898c779309ece2ef8dc81f635900545c/src/angular-app/bellows/core/api/json-rpc.service.ts#L148)):*
> ```XML
> 
> export const JsonRpcModule = angular
>   .module('jsonRpc', [ErrorModule, ExceptionOverrideModule])
> ```

`ExceptionOverrideModule' does not need to be a dependency here since it is already a dependency of  `CoreModule` and `JsonRpcModule` is a sister dependency.

---


*Comments from [Reviewable](https://reviewable.io/reviews/sillsdev/web-languageforge/227)*
<!-- Sent from Reviewable.io -->

ermshiperete commented:
Review status: 12 of 19 files reviewed at latest revision, 9 unresolved discussions.

---

*[src/angular-app/bellows/core/exception-handling.service.ts, line 3 at r1](https://beta.reviewable.io/reviews/sillsdev/web-languageforge/227#-LAuf5Hs2xzHJbl9IDKb:-LB1ABke-z2JeGSPKtPs:b-896fix) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/angular-app/bellows/core/exception-handling.service.ts#L3)):*
<details><summary><i>Previously, irahopkinson (Ira Hopkinson) wrote…</i></summary><blockquote>

Umm, `Client` is vendor code, move the blank line down one more.
</blockquote></details>

Done.

---

*[src/angular-app/bellows/core/exception-handling.service.ts, line 66 at r1](https://beta.reviewable.io/reviews/sillsdev/web-languageforge/227#-LAuiCmJ7r1Q6j1foBnE:-LB1KmzR8xqHqw5x3Ip9:b-896fix) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/angular-app/bellows/core/exception-handling.service.ts#L66)):*
<details><summary><i>Previously, irahopkinson (Ira Hopkinson) wrote…</i></summary><blockquote>

The entire `.factory(...)` line that was quoted.
</blockquote></details>

Done.

---

*[src/angular-app/bellows/core/exception-handling.service.ts, line 61 at r2](https://beta.reviewable.io/reviews/sillsdev/web-languageforge/227#-LB-DH-uCQxGluIcmBZK:-LB-sEov3IScMyDVlvkz:biivgq1) ([raw file](https://github.com/sillsdev/web-languageforge/blob/848b3151898c779309ece2ef8dc81f635900545c/src/angular-app/bellows/core/exception-handling.service.ts#L61)):*
<details><summary><i>Previously, irahopkinson (Ira Hopkinson) wrote…</i></summary><blockquote>

This service is not doing anything and can be removed (since you are using the factory below).
</blockquote></details>

If I remove it I get an error: `Error: [$injector:unpr] Unknown provider: exceptionHandlerProvider <- exceptionHandler <- jsonRpc <- apiService <- projectService` since `ExceptionHandlingService` is instantiated in the factory.

But it turns out I have to remove the `new ExceptionHandlingService` below.

---

*[src/angular-app/bellows/core/api/json-rpc.service.ts, line 29 at r2](https://beta.reviewable.io/reviews/sillsdev/web-languageforge/227#-LB-0vdhC3F6gZWhfGll:-LB1JqAJCTYBWyXA0PqG:bmmoft4) ([raw file](https://github.com/sillsdev/web-languageforge/blob/848b3151898c779309ece2ef8dc81f635900545c/src/angular-app/bellows/core/api/json-rpc.service.ts#L29)):*
<details><summary><i>Previously, irahopkinson (Ira Hopkinson) wrote…</i></summary><blockquote>

I'm wondering why the `ExceptionHandlingService` is injected here since it is never used? Pretty sure it can be removed.
</blockquote></details>

It's used now :smiley: 

---

*[src/angular-app/bellows/core/api/json-rpc.service.ts, line 148 at r2](https://beta.reviewable.io/reviews/sillsdev/web-languageforge/227#-LB-DvVW6IGxUREYGCgt:-LB1KGIoA2kq0ZVFmrd3:b-896fix) ([raw file](https://github.com/sillsdev/web-languageforge/blob/848b3151898c779309ece2ef8dc81f635900545c/src/angular-app/bellows/core/api/json-rpc.service.ts#L148)):*
<details><summary><i>Previously, irahopkinson (Ira Hopkinson) wrote…</i></summary><blockquote>

`ExceptionOverrideModule' does not need to be a dependency here since it is already a dependency of  `CoreModule` and `JsonRpcModule` is a sister dependency.
</blockquote></details>

Done.

---

*[src/Api/Library/Shared/Palaso/Exception/BugsnagExceptionHandler.php, line 12 at r1](https://beta.reviewable.io/reviews/sillsdev/web-languageforge/227#-LAunQaZE3jmcemOOO-v:-LB1KLso3u5iobks2BIz:b-2kdt1r) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/Api/Library/Shared/Palaso/Exception/BugsnagExceptionHandler.php#L12)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Done.

But then I'm missing something else. With these three lines (plus the `$appPath =  realpath...`) removed, I get an error:

```
Notice: Use of undefined constant VERSION - assumed 'VERSION' in 
languageforge/src/Api/Library/Shared/Palaso/Exception/BugsnagExceptionHandler.php on line 24
Call Stack
#	Time	Memory	Function	Location
1	0.0074	361232	{main}( )	.../index.php:0
2	0.0654	395112	Api\Library\Shared\Palaso\Exception\BugsnagExceptionHandler::setup( ???, ??? )	.../index.php:26
```

So it seems I need the `require_once $appPath . '/version.php';` line.
</blockquote></details>

This has been moved to `index.php`.

---

*[src/Api/Library/Shared/Palaso/Exception/BugsnagExceptionHandler.php, line 44 at r1](https://beta.reviewable.io/reviews/sillsdev/web-languageforge/227#-LAv2zQ60auErG4gID5t:-LB1Kgdk-9Cx6htHXxFb:b-896fix) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/Api/Library/Shared/Palaso/Exception/BugsnagExceptionHandler.php#L44)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Yes. I noticed that the projectId isn't very useful when I looked at the entries that got created on Bugsnag.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/sillsdev/web-languageforge/227)*
<!-- Sent from Reviewable.io -->

ermshiperete commented:
Review status: 12 of 19 files reviewed at latest revision, 9 unresolved discussions.

---

*[src/angular-app/bellows/core/exception-handling.service.ts, line 19 at r3](https://beta.reviewable.io/reviews/sillsdev/web-languageforge/227#-LB1L23yAN696PyeLvDI:-LB1L23yAN696PyeLvDJ:b-yeqfzc) ([raw file](https://github.com/sillsdev/web-languageforge/blob/dc642a02f3016e05a5f486f4a4787e84fae6c23b/src/angular-app/bellows/core/exception-handling.service.ts#L19)):*
> ```
> '0ac621681992952c356d54bc01d529b1'
> ```

This is still WIP.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/sillsdev/web-languageforge/227)*
<!-- Sent from Reviewable.io -->

ermshiperete commented:
Review status: 12 of 20 files reviewed at latest revision, 10 unresolved discussions.

---

*[src/angular-app/bellows/core/exception-handling.service.ts, line 19 at r3](https://beta.reviewable.io/reviews/sillsdev/web-languageforge/227#-LB1L23yAN696PyeLvDI:-LB1mTBBEiksv9DvjRNI:b-l80y73) ([raw file](https://github.com/sillsdev/web-languageforge/blob/dc642a02f3016e05a5f486f4a4787e84fae6c23b/src/angular-app/bellows/core/exception-handling.service.ts#L19)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

> ```
> '0ac621681992952c356d54bc01d529b1'
> ```

This is still WIP.
</blockquote></details>

No longer WIP :smiley: 

I suggest we leave a default key that points to a different project that is not linked to JIRA or Trello. This would allow to catch exception that happen before we have a session object.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/sillsdev/web-languageforge/227)*
<!-- Sent from Reviewable.io -->

irahopkinson commented:
Reviewed 5 of 7 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[src/angular-app/bellows/core/api/json-rpc.service.ts, line 77 at r4](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LB3W4kO8oQAUo432qy0:-LB3W4kO8oQAUo432qy1:b-mbvxii) ([raw file](https://github.com/sillsdev/web-languageforge/blob/81108d281e92cd3f9a069c0ca13e5c5185d700f8/src/angular-app/bellows/core/api/json-rpc.service.ts#L77)):*
> ```XML
>       if (response.data === null) {
>         // TODO error handling for jsonRpc CP 2013-07
>         this.error.notify('RPC Error', 'data is null');
> ```

I'm wondering if these are changed to notify instead of error, I think our E2E tests won't pick them up as failures anymore. Not sure if that is important or not?

---


*Comments from [Reviewable](https://reviewable.io/reviews/sillsdev/web-languageforge/227)*
<!-- Sent from Reviewable.io -->

ermshiperete commented:
Review status: 10 of 18 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/angular-app/bellows/core/api/json-rpc.service.ts, line 77 at r4](https://beta.reviewable.io/reviews/sillsdev/web-languageforge/227#-LB3W4kO8oQAUo432qy0:-LB6hdvY9TxVSkrERidu:b-6bfxjg) ([raw file](https://github.com/sillsdev/web-languageforge/blob/81108d281e92cd3f9a069c0ca13e5c5185d700f8/src/angular-app/bellows/core/api/json-rpc.service.ts#L77)):*
<details><summary><i>Previously, irahopkinson (Ira Hopkinson) wrote…</i></summary><blockquote>

I'm wondering if these are changed to notify instead of error, I think our E2E tests won't pick them up as failures anymore. Not sure if that is important or not?
</blockquote></details>

I added the `notify()` method and changed to that so that we don't display the dialog that says "Please send us an email that this happened" since the error now gets automatically send to bugsnag.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/sillsdev/web-languageforge/227)*
<!-- Sent from Reviewable.io -->

irahopkinson commented:
Reviewed 8 of 10 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[src/angular-app/bellows/core/exception-handling.service.ts, line 27 at r5](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LBY2tBSBpzAf835nino:-LBY2tBSBpzAf835ninp:b3cz5ae) ([raw file](https://github.com/sillsdev/web-languageforge/blob/52f4902ea047805923b9d89a23a402ea0891a25b/src/angular-app/bellows/core/exception-handling.service.ts#L27)):*
> ```XML
>   }
> 
>   private notifyBugsnag(exception: Error, cause?: string) {
> ```

Style: private methods are listed after public ones

---

*[src/Api/Model/Shared/Command/SessionCommands.php, line 76 at r5](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LBY2VE9-SD0TdrSF-8T:-LBY2VE9-SD0TdrSF-8U:b3cz5ae) ([raw file](https://github.com/sillsdev/web-languageforge/blob/52f4902ea047805923b9d89a23a402ea0891a25b/src/Api/Model/Shared/Command/SessionCommands.php#L76)):*
> ```PHP
>     }
> 
>     private static function getSessionDirectory($mockFilename = null)
> ```

Style: private methods are listed after public ones

---


*Comments from [Reviewable](https://reviewable.io/reviews/sillsdev/web-languageforge/227)*
<!-- Sent from Reviewable.io -->

irahopkinson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Great work learning AngularJS, TypeScript and PHP. I added a couple of style comments but no need to hold this up.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-:-LBY3guo--u9BXTzYDgZ:b-tcmmt7)*
<!-- Sent from Reviewable.io -->

megahirt commented:
I am satisfied with this PR and will be merging it today!  @ermshiperete as a follow-up, separate PR, I would like all bugsnag client implementations (C#, PHP, AngularJS) maintain a list of production domains.  If we're not on a production domain, then ignore the error and don't notify bugsnag.

---

Reviewed 3 of 14 files at r1, 5 of 9 files at r2, 1 of 7 files at r3, 1 of 3 files at r4, 8 of 10 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[src/angular-app/bellows/core/exception-handling.service.ts, line 15 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LAuhUZHEbZD2oD_7dtb:-LBZrCdNCP4yPvlXvDkk:b162r3p) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/angular-app/bellows/core/exception-handling.service.ts#L15)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Not finished yet :smile: 
</blockquote></details>

This apiKey should probably also be stamped at build time, but we don't have a precedent for that I think in JS.  Maybe we should have a separate JSON file for these secrets that lives outside our web space?  Just thinking out loud here.

---

*[src/angular-app/bellows/core/exception-handling.service.ts, line 28 at r5](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LBZs25R5782Y3GAa9EL:-LBZs25R5782Y3GAa9EM:bz0b24g) ([raw file](https://github.com/sillsdev/web-languageforge/blob/52f4902ea047805923b9d89a23a402ea0891a25b/src/angular-app/bellows/core/exception-handling.service.ts#L28)):*
> ```XML
> 
>   private notifyBugsnag(exception: Error, cause?: string) {
>     if (this.metadata.userName.startsWith('test_runner_')) {
> ```

OK this is good to know.  @ermshiperete I remember you saying that you needed to figure out how to not fire off bugsnag events for running e2e tests.  We really should go farther than this though.  Ideally we should only produce bugsnag events for production sites.  So only if the domain matches one of our production domains, specifically "scriptureforge.org", "jamaicanpsalms.scriptureforge.org",  "languageforge.org" and "waaqwiinaagiwritings.org"  In the future we should also have "beta.languageforge.org", "beta.scriptureforge.org", "demo.scriptureforge.org" and "demo.languageforge.org" as well, but those don't exist yet.  Basically, if you're not on one of these sites, bugsnag should be disabled.

---


*Comments from [Reviewable](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-:-LBZtNpA4cTGb7RuIYYc:bchwhcs)*
<!-- Sent from Reviewable.io -->

megahirt commented:
I will merge this as soon as the reviewable discussions are resolved.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-:-LBZu2BVCdsOvA5JrcVd:b79ppt1)*
<!-- Sent from Reviewable.io -->

ermshiperete commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[src/angular-app/bellows/core/exception-handling.service.ts, line 15 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LAuhUZHEbZD2oD_7dtb:-LB_OOjcAQgTJPtkF4F7:bej91lp) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5234f32a4f0ab816d24d9ec2ed82e18581d58de2/src/angular-app/bellows/core/exception-handling.service.ts#L15)):*
<details><summary><i>Previously, megahirt (Chris Hirt) wrote…</i></summary><blockquote>

This apiKey should probably also be stamped at build time, but we don't have a precedent for that I think in JS.  Maybe we should have a separate JSON file for these secrets that lives outside our web space?  Just thinking out loud here.
</blockquote></details>

I agree. I'll see if I can find a way to do that.

---

*[src/angular-app/bellows/core/exception-handling.service.ts, line 28 at r5](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LBZs25R5782Y3GAa9EL:-LB_OgKwE31cpGzj1R8I:b-46ospx) ([raw file](https://github.com/sillsdev/web-languageforge/blob/52f4902ea047805923b9d89a23a402ea0891a25b/src/angular-app/bellows/core/exception-handling.service.ts#L28)):*
<details><summary><i>Previously, megahirt (Chris Hirt) wrote…</i></summary><blockquote>

OK this is good to know.  @ermshiperete I remember you saying that you needed to figure out how to not fire off bugsnag events for running e2e tests.  We really should go farther than this though.  Ideally we should only produce bugsnag events for production sites.  So only if the domain matches one of our production domains, specifically "scriptureforge.org", "jamaicanpsalms.scriptureforge.org",  "languageforge.org" and "waaqwiinaagiwritings.org"  In the future we should also have "beta.languageforge.org", "beta.scriptureforge.org", "demo.scriptureforge.org" and "demo.languageforge.org" as well, but those don't exist yet.  Basically, if you're not on one of these sites, bugsnag should be disabled.
</blockquote></details>

ok, I'll add that.

---


*Comments from [Reviewable](https://reviewable.io/reviews/sillsdev/web-languageforge/227)*
<!-- Sent from Reviewable.io -->

ermshiperete commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[src/Api/Model/Shared/Command/SessionCommands.php, line 76 at r5](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LBY2VE9-SD0TdrSF-8T:-LB_OtAn3CdoSBvoU5_c:b-d7bhn0) ([raw file](https://github.com/sillsdev/web-languageforge/blob/52f4902ea047805923b9d89a23a402ea0891a25b/src/Api/Model/Shared/Command/SessionCommands.php#L76)):*
<details><summary><i>Previously, irahopkinson (Ira Hopkinson) wrote…</i></summary><blockquote>

Style: private methods are listed after public ones
</blockquote></details>

will do

---


*Comments from [Reviewable](https://reviewable.io/reviews/sillsdev/web-languageforge/227)*
<!-- Sent from Reviewable.io -->

irahopkinson commented:
Reviewed 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/sillsdev/web-languageforge/227)*
<!-- Sent from Reviewable.io -->

ermshiperete commented:
Review status: 16 of 18 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/angular-app/bellows/core/exception-handling.service.ts, line 27 at r5](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LBY2tBSBpzAf835nino:-LBajuP8AZZgc7MaMzRF:b-896fix) ([raw file](https://github.com/sillsdev/web-languageforge/blob/52f4902ea047805923b9d89a23a402ea0891a25b/src/angular-app/bellows/core/exception-handling.service.ts#L27)):*
<details><summary><i>Previously, irahopkinson (Ira Hopkinson) wrote…</i></summary><blockquote>

Style: private methods are listed after public ones
</blockquote></details>

Done.

---

*[src/Api/Model/Shared/Command/SessionCommands.php, line 76 at r5](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LBY2VE9-SD0TdrSF-8T:-LBajviQCh0XC7ikL6Vf:b-896fix) ([raw file](https://github.com/sillsdev/web-languageforge/blob/52f4902ea047805923b9d89a23a402ea0891a25b/src/Api/Model/Shared/Command/SessionCommands.php#L76)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

will do
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/sillsdev/web-languageforge/227)*
<!-- Sent from Reviewable.io -->

irahopkinson commented:
Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/sillsdev/web-languageforge/227)*
<!-- Sent from Reviewable.io -->

ermshiperete commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[src/angular-app/bellows/core/exception-handling.service.ts, line 28 at r5](https://reviewable.io/reviews/sillsdev/web-languageforge/227#-LBZs25R5782Y3GAa9EL:-LBfS94R0Q1mG0x9NyvT:b-gugtfw) ([raw file](https://github.com/sillsdev/web-languageforge/blob/52f4902ea047805923b9d89a23a402ea0891a25b/src/angular-app/bellows/core/exception-handling.service.ts#L28)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

ok, I'll add that.
</blockquote></details>

I think we should also enable bugsnag for qa: if we get an exception while testing on qa we want to know.

---


*Comments from [Reviewable](https://reviewable.io/reviews/sillsdev/web-languageforge/227)*
<!-- Sent from Reviewable.io -->
